File: Sources/fXDKit/Classes/_ObjC/_module/FXDmoduleCloud.m


#import "FXDmoduleCloud.h"


@implementation FXDmoduleCloud

#pragma mark - Memory management
- (void)dealloc {
	_statusCallback = nil;
}

#pragma mark - Initialization

#pragma mark - Property overriding

#pragma mark - Method overriding

#pragma mark - Public
- (void)prepareContainerURLwithIdentifier:(NSString*)containerIdentifier withStatusCallback:(FXDcallbackFinish)statusCallback {	FXDLog_DEFAULT

	FXDLogObject(containerIdentifier);
	FXDLogObject(statusCallback);

	_containerIdentifier = containerIdentifier;
	_statusCallback = statusCallback;


	[self observedNSUbiquityIdentityDidChange:nil];
}

#pragma mark -
- (void)notifyCallbackWithContainerURL:(NSURL*)containerURL shouldAddObserver:(BOOL)shouldAddObserver withAlertBody:(NSString*)alertBody {	FXDLog_DEFAULT

	//MARK: Assume if notification is nil, observer should be added
	if (shouldAddObserver) {
		NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];

		[notificationCenter
		 addObserver:self
		 selector:@selector(observedNSUbiquityIdentityDidChange:)
		 name:NSUbiquityIdentityDidChangeNotification
		 object:nil];
	}

	if (alertBody.length > 0) {

		//TODO: To be used with Swift version of simpleAlert
		UIAlertController *alertController = [UIAlertController
											  alertControllerWithTitle:alertBody
											  message:nil
											  preferredStyle:UIAlertControllerStyleAlert];

		UIAlertAction *cancelAction =
		[UIAlertAction
		 actionWithTitle:NSLocalizedString(@"OK", nil)
		 style:UIAlertActionStyleCancel
		 handler:nil];

		[alertController addAction:cancelAction];


		UIWindow *mainWindow = [[UIApplication sharedApplication] performSelector:@selector(mainWindow:)];
		FXDLogObject(mainWindow);

		UIViewController *presentingScene = mainWindow.rootViewController;
		FXDLogObject(presentingScene);

		[presentingScene
		 presentViewController:alertController
		 animated:YES
		 completion:nil];
	}

	if (_statusCallback) {
		_statusCallback(_cmd, (containerURL != nil), containerURL);
	}
}


#pragma mark - Observer
- (void)observedNSUbiquityIdentityDidChange:(NSNotification*)notification {	FXDLog_DEFAULT
	FXDLogObject(notification);

	NSFileManager *fileManager = [NSFileManager defaultManager];
	FXDLogObject(fileManager.ubiquityIdentityToken);

	if (fileManager.ubiquityIdentityToken == nil) {
		[self
		 notifyCallbackWithContainerURL:nil
		 shouldAddObserver:(notification == nil)
		 withAlertBody:NSLocalizedString(@"Please enable iCloud", nil)];
		return;
	}


	NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];

	NSError *error = nil;
	id<NSObject,NSCopying,NSCoding> identityToken = [NSKeyedUnarchiver unarchivedObjectOfClass:[NSObject class] fromData:[userDefaults objectForKey:userdefaultObjSavedUbiquityIdentityToken] error:&error];
	FXDLogObject(identityToken);
	FXDLogBOOL([identityToken isEqual:fileManager.ubiquityIdentityToken]);
	FXDLog_ERROR

	if (identityToken == nil
		|| [identityToken isEqual:fileManager.ubiquityIdentityToken] == NO) {

		identityToken = fileManager.ubiquityIdentityToken;

		NSData *archivedData = [NSKeyedArchiver archivedDataWithRootObject:identityToken requiringSecureCoding:NO error:&error];
		FXDLog_ERROR

		[userDefaults setObject:archivedData forKey:userdefaultObjSavedUbiquityIdentityToken];
		[userDefaults synchronize];
	}


	self.containerURL = [NSURL evaluatedURLforPath:[userDefaults objectForKey:userdefaultStringSavedUbiquityContainerURL]];
	FXDLogObject(self.containerURL);


	NSOperationQueue *containerURLupdatingQueue = [NSOperationQueue newSerialQueueWithName:NSStringFromSelector(_cmd)];

	[containerURLupdatingQueue
	 addOperationWithBlock:^{	FXDLog_DEFAULT
		 FXDLogObject(self.containerIdentifier);

		 NSURL *ubiquityContainerURL = [fileManager URLForUbiquityContainerIdentifier:self.containerIdentifier];
		 FXDLogObject(ubiquityContainerURL);


		 [[NSOperationQueue mainQueue]
		  addOperationWithBlock:^{	FXDLog_DEFAULT
			  FXDLogObject(self.containerURL);

			  FXDLogBOOL([[ubiquityContainerURL absoluteString] isEqualToString:[self.containerURL absoluteString]]);

			  if (ubiquityContainerURL == nil && self.containerURL == nil) {
				  [self
				   notifyCallbackWithContainerURL:nil
				   shouldAddObserver:(notification == nil)
				   withAlertBody:NSLocalizedString(@"iCloud cannot be activated currently", nil)];
				  return;
			  }


			  [userDefaults setObject:ubiquityContainerURL.absoluteString forKey:userdefaultStringSavedUbiquityContainerURL];
			  [userDefaults synchronize];


			  self.containerURL = ubiquityContainerURL;

			  FXDLogObject([fileManager infoDictionaryForFolderURL:self.containerURL]);

			  [self
			   notifyCallbackWithContainerURL:self.containerURL
			   shouldAddObserver:(notification == nil)
			   withAlertBody:nil];
		  }];
	 }];
}

#pragma mark - Delegate

@end

File: Sources/FXDmoduleCloud.swift
import Foundation
import UIKit

let userdefaultObjSavedUbiquityIdentityToken = "SavedUbiquityIdentityTokenObjKey"
let userdefaultStringSavedUbiquityContainerURL = "SavedUbiquityContainerURLstringKey"

class FXDmoduleCloud: NSObject, @unchecked Sendable {
	var statusCallback: FXDcallback?

	var containerIdentifier: String?
	var containerURL: URL?

	deinit {
		statusCallback = nil
	}

    @MainActor open func prepareContainerURLwithIdentifier(_ containerIdentifier: String?, withStatusCallback statusCallback: FXDcallback?) {	fxd_log()
		fxdPrint("containerIdentifier: ", containerIdentifier)
		fxdPrint("statusCallback: ", statusCallback)

		self.containerIdentifier = containerIdentifier
		self.statusCallback = statusCallback

		let notificationCenter = NotificationCenter.default
		notificationCenter.addObserver(
			self,
			selector: #selector(observedNSUbiquityIdentityDidChange(_:)),
			name: NSNotification.Name.NSUbiquityIdentityDidChange,
			object: nil)

		observedNSUbiquityIdentityDidChange(nil)
	}

    @MainActor open func notifyCallback(withContainerURL containerURL: URL?, withAlertBody alertBody: String?) {	fxd_log()
		if alertBody?.isEmpty ?? false {
			UIAlertController.simpleAlert(
				withTitle: alertBody,
				message: nil)
		}

		statusCallback?((containerURL != nil), containerURL)
	}

    @MainActor @objc open func observedNSUbiquityIdentityDidChange(_ notification: Notification?) {	fxd_log()
		fxdPrint("notification: ", notification)

		fxdPrint("FileManager.default.ubiquityIdentityToken: ", FileManager.default.ubiquityIdentityToken)

		guard FileManager.default.ubiquityIdentityToken != nil else {
			notifyCallback(
				withContainerURL: nil,
				withAlertBody: NSLocalizedString("Please enable iCloud", comment: ""))
			return
		}

		let savedIdentityTokenData = UserDefaults.standard.object(forKey: userdefaultObjSavedUbiquityIdentityToken) as? Data

		var identityToken: (NSCoding & NSCopying & NSObjectProtocol)?
		do {
			identityToken = try NSKeyedUnarchiver.unarchivedObject(ofClasses: [NSCoding.self, NSCopying.self, NSObjectProtocol.self], from: savedIdentityTokenData!) as? any NSCoding & NSCopying & NSObjectProtocol
		} catch {
			fxdPrint("\(error)")
		}

		fxdPrint("identityToken: ", identityToken)
		fxdPrint("identityToken?.isEqual(fileManager.ubiquityIdentityToken): ", identityToken?.isEqual(FileManager.default.ubiquityIdentityToken))

		if identityToken == nil
			|| identityToken!.isEqual(FileManager.default.ubiquityIdentityToken) == false {

			do {
				let defaultIdentityToken: Any = FileManager.default.ubiquityIdentityToken as Any
				let defaultIdentityTokenData = try NSKeyedArchiver.archivedData(withRootObject: defaultIdentityToken, requiringSecureCoding: false)

				UserDefaults.standard.setValue(defaultIdentityTokenData, forKey: userdefaultObjSavedUbiquityIdentityToken)
				UserDefaults.standard.synchronize()
			} catch {
				fxdPrint("\(error)")
			}
		}

		let savedContainerURL = UserDefaults.standard.object(forKey: userdefaultStringSavedUbiquityContainerURL) as? String
		containerURL = NSURL.evaluatedURLforPath(savedContainerURL)
		fxdPrint("containerURL: ", containerURL)

		let containerURLupdatingQueue = OperationQueue.newSerialQueue(withName: #function)

		containerURLupdatingQueue?.addOperation {	fxd_log()
			fxdPrint("containerIdentifier: ", self.containerIdentifier)

			let ubiquityContainerURL = FileManager.default.url(forUbiquityContainerIdentifier: self.containerIdentifier)
			fxdPrint("ubiquityContainerURL: ", ubiquityContainerURL)

			DispatchQueue.main.async {	fxd_log()
				guard ubiquityContainerURL != nil
						|| self.containerURL != nil
				else {
					self.notifyCallback(
						withContainerURL: nil,
						withAlertBody: NSLocalizedString("iCloud cannot be activated currently", comment: ""))
					return
				}

				UserDefaults.standard.setValue(ubiquityContainerURL?.absoluteString, forKey: userdefaultStringSavedUbiquityContainerURL)
				UserDefaults.standard.synchronize()

				self.containerURL = ubiquityContainerURL
				fxdPrint("", FileManager.default.infoDictionary(forFolderURL: self.containerURL))

				self.notifyCallback(
					withContainerURL: self.containerURL,
					withAlertBody: nil)
			}
		}
	}
}
